x-pack/solutions/observability/plugins/apm/common/service_map/get_service_map_nodes.ts

```diff
@@ -0,0 +1,292 @@
+
+        map.set(node.id, {
+          ...serviceNode,
+          ...(serviceAnomalyStats ? { serviceAnomalyStats } : null),
+        });
+      }
+    } else {
+      const exitSpanNodes = exitSpans.get(id) ?? [];
+      if (exitSpanNodes.length > 0) {
+        const exitSpanSample = exitSpanNodes[0];
+        map.set(id, {
+          ...exitSpanSample,
+          label: exitSpanSample[SPAN_DESTINATION_SERVICE_RESOURCE],
+          [SPAN_TYPE]: exitSpanNodes.map((n) => n[SPAN_TYPE]).sort()[0],
+          [SPAN_SUBTYPE]: exitSpanNodes.map((n) => n[SPAN_SUBTYPE]).sort()[0],
+        });
+      }
+    }
+
+    return map;
+  }, new Map<string, ConnectionNode>());
+}
+
+function mapEdges({
+  allConnections,
+  nodes,
+}: {
+  allConnections: Connection[];
+  nodes: Map<string, ConnectionNode>;
+}) {
+  const connections = allConnections.reduce((acc, connection) => {
+    const sourceData = nodes.get(connection.source.id);
+    const targetData = nodes.get(connection.destination.id);
+
+    if (!sourceData || !targetData || sourceData.id === targetData.id) {
+      return acc;
+    }
+
+    const label = `${sourceData[SERVICE_NAME]} to ${
+      targetData[SERVICE_NAME] || targetData[SPAN_DESTINATION_SERVICE_RESOURCE]
+    }`;
+    const id = getConnectionId({ source: sourceData, destination: targetData });
+
+    acc.set(id, {
+      source: sourceData.id,
+      target: targetData.id,
+      label,
+      id,
+      sourceData,
+      targetData,
+    });
+
+    return acc;
+  }, new Map<string, ConnectionEdge & { sourceData: ConnectionNode; targetData: ConnectionNode }>());
+
+  return Array.from(connections.values());
+}
+
+function markBidirectionalConnections({ connections }: { connections: ConnectionEdge[] }) {
+  return connections.reduce<ConnectionEdge[]>((prev, connection) => {
+    const reversedConnection = prev.find(
+      (c) => c.target === connection.source && c.source === connection.target
+    );
+
+    if (reversedConnection) {
+      reversedConnection.bidirectional = true;
+      connection.isInverseEdge = true;
+    }
+
+    return prev.concat(connection);
+  }, []);
+}
+
+export function getServiceMapNodes({
+  connections,
+  exitSpanDestinations,
+  servicesData,
+  anomalies,
+}: ServiceMapConnections): GroupResourceNodesResponse {
+  const allConnections = addMessagingConnections(connections, exitSpanDestinations);
+  const allNodes = getAllNodes(servicesData, allConnections);
+  const allServices = getAllServices(allNodes, exitSpanDestinations);
+
+  const nodes = mapNodes({
+    allConnections,
+    nodes: allNodes,
+    services: allServices,
+    exitSpanDestinations,
+    anomalies,
+  });
+
+  // Build connections with mapped nodes
+  const mappedEdges = mapEdges({ allConnections, nodes });
+
+  const uniqueNodes = mappedEdges
+    .flatMap((connection) => [connection.sourceData, connection.targetData])
+    .concat(Array.from(allServices.values()))
```

Comment:
```suggestion
    .concat(...allServices.values())
```
We can just spread instead of doing an additional array allocation